File: VindsidenApp/RHCStationCell.m
//
//  RHCStationCell.m
//  Haugastol-v2
//
//  Created by Ragnar Henriksen on 25.04.13.
//  Copyright (c) 2013 RHC. All rights reserved.
//

@import VindsidenKit;

#import "RHCStationCell.h"
#import "NSSet+Sort.h"

#import "NSObject+performBlockCancel.h"
#import "RHEGraphView.h"


@interface RHCStationCell ()

@property (strong, nonatomic) NSDateFormatter *dateFormatter;

@property (strong, nonatomic) NSTimer *updatedTimer;

@end


@implementation RHCStationCell

@synthesize currentStation = _currentStation;


- (void)awakeFromNib
{
    [super awakeFromNib];

    self.updatedAtLabel.text = @"";
    self.cameraButton.alpha = 0.0;
}


- (void)prepareForReuse
{
    [super prepareForReuse];

    [self.updatedTimer invalidate];
    self.updatedTimer = nil;

    self.updatedAtLabel.text = NSLocalizedString(@"LABEL_UPDATING", @"Updating");

    self.cameraButton.alpha = 0.0;
    self.graphView.plots = nil;
    [self.stationView resetInfoLabels];
}


- (NSDateFormatter *) dateFormatter
{
    if ( _dateFormatter ) {
        return _dateFormatter;
    }

    _dateFormatter = [[NSDateFormatter alloc] init];
    [_dateFormatter setDateStyle:NSDateFormatterMediumStyle];
    [_dateFormatter setTimeStyle:NSDateFormatterMediumStyle];

    return _dateFormatter;
}


- (void)updateLastUpdatedLabel
{
    CDPlot *plot = self.currentStation.lastRegisteredPlot;

    if ( plot != nil ) {
        if ( [plot.plotTime compare:[NSDate date]] == NSOrderedAscending ) {
            self.updatedAtLabel.text = [[AppConfig sharedConfiguration] relativeDate:plot.plotTime];
        } else {
            self.updatedAtLabel.text = [[AppConfig sharedConfiguration] relativeDate:nil];
        }
    } else {
        self.updatedAtLabel.text = NSLocalizedString(@"LABEL_NOT_UPDATED", @"Not updated");
    }
}


- (void)displayPlots
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [self syncDisplayPlots];
    });
}


- (void)syncDisplayPlots
{
    NSDate *inDate = [[NSDate date] dateByAddingTimeInterval:-1*(kPlotHistoryHours-1)*3600];
    NSCalendar *gregorian = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
    NSDateComponents *inputComponents = [gregorian components:(NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay | NSCalendarUnitHour) fromDate:inDate];
    NSDate *outDate = [gregorian dateFromComponents:inputComponents];
    NSManagedObjectContext *context = self.currentStation.managedObjectContext;
    NSFetchRequest *fetchRequest = CDPlot.fetchRequest;

    fetchRequest.predicate = [NSPredicate predicateWithFormat:@"station == %@ AND plotTime >= %@", self.currentStation, outDate];
    fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"plotTime" ascending:NO]];

    NSArray *cdplots = [context executeFetchRequest:fetchRequest error:nil];

    if ( [cdplots count] ) {
        self.graphView.plots = cdplots;
        [self.stationView updateWithPlot:cdplots.firstObject];
        if ( [[cdplots[0] plotTime] compare:[NSDate date]] == NSOrderedAscending ) {
            self.updatedAtLabel.text = [[AppConfig sharedConfiguration] relativeDate:[cdplots.firstObject plotTime]];
        } else {
            self.updatedAtLabel.text = [[AppConfig sharedConfiguration] relativeDate:nil];
        }
    } else {
        self.updatedAtLabel.text = NSLocalizedString(@"LABEL_NOT_UPDATED", @"Not updated");
    }
}


#pragma mark -


- (void)setCurrentStation:(CDStation *)currentStation
{
    _currentStation = currentStation;

    self.stationNameLabel.text = self.currentStation.stationName;
    [self displayPlots];

    [self.updatedTimer invalidate];
    self.updatedTimer = [[NSTimer alloc] initWithFireDate:[NSDate date]
                                                 interval:1
                                                   target:self
                                                 selector:@selector(updateLastUpdatedLabel)
                                                 userInfo:nil
                                                  repeats:YES];
    [[NSRunLoop currentRunLoop] addTimer:self.updatedTimer forMode:NSDefaultRunLoopMode];
}


@end

File: VindsidenApp/RHCStationCell.swift
//
//  RHCStationCell.swift
//  VindsidenApp
//
//  Created by Ragnar Henriksen on 12/10/2023.
//  Copyright © 2023 RHC. All rights reserved.
//

import UIKit
import OSLog
import SwiftUI
import VindsidenKit
import Units

@objc
class RHCStationCell: UICollectionViewCell {
    @ObservedObject var observer: PlotObservable = PlotObservable()

    override init(frame: CGRect) {
        fatalError("Not implemented")
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)

        setupView()
    }

    func setupView() {
//        @ObservedObject
//        var settings = (UIApplication.shared.delegate as? RHCAppDelegate)!.settings

        contentConfiguration = UIHostingConfiguration(content: {
            StationView(observer: observer, updater: displayPlots)
//                .environmentObject(settings)
                .environment(\.managedObjectContext, DataManager.shared.viewContext())
        })
    }

    @objc
    weak var currentStation: CDStation? {
        didSet {
            guard let station = currentStation else {
                return
            }

            observer.station = station

            displayPlots()
        }
    }

    @objc
    func displayPlots() {
        DispatchQueue.main.async { [weak self] in
            self?.syncDisplayPlots()
        }
    }

    private func syncDisplayPlots() {
        guard
            let gregorian = NSCalendar(identifier: .gregorian),
            let currentStation,
            let context = currentStation.managedObjectContext
        else {
            return
        }

        let inDate = Date().addingTimeInterval(-1*AppConfig.Global.plotHistory*3600)
        let inputComponents = gregorian.components([.year, .month, .day, .hour], from: inDate)
        let outDate = gregorian.date(from: inputComponents) ?? Date()

        let fetchRequest = CDPlot.fetchRequest()

        fetchRequest.predicate = NSPredicate(format: "station == %@ AND plotTime >= %@", currentStation, outDate as CVarArg)
        fetchRequest.fetchLimit = 35
        fetchRequest.sortDescriptors = [
            NSSortDescriptor(key: "plotTime", ascending: false),
        ]

        let cdplots: [CDPlot] = (try? context.fetch(fetchRequest)) ?? []

        observer.plot = cdplots.first
        observer.plots = cdplots
    }
}
